### PR TITLE
Danik usenko

### PR DESCRIPTION
**Pull request by danikusenko**
add courseViemModel and validation of Modules, components, lessons

### CODE REVIEW

File: Data/DataRepository.cs [hunk 1]
```diff
@@ -174,23 +196,203 @@ public Comment GetComment(int commentId)
...
+
+        public void InsertToCategoryCourse(int courseId, int categoryId)
+        {
+            using (var connection = new MySqlConnection(_connectionString))
+            {
+                connection.Open();
+                connection.Execute(
+                    @"INSERT INTO category_course(course_id, category_id) 
+                      VALUES (@CourseId, @CategoryId);",
+                    new { CourseId =courseId, CategotyId = categoryId }
```
Comment (kravien) [inline Data/DataRepository.cs]: There is `CategoryId` typo.

File: Data/DataRepository.cs [hunk 2]
```diff
@@ -174,23 +196,203 @@ public Comment GetComment(int commentId)
...
+                        Title = course.Title,
+                        Description = course.Description,
+                        Price = course.Price,
+                        AcceptancePercantage = course.AcceptancePercantage,
+                        Requirements = course.Requirements,
+                        Status = status,
+                        PublishedAt = DateTime.Now
+                    }
+                ));
+                course.Id = courseId;
```
Comment (kravien) [inline Data/DataRepository.cs:310]: Now I see what you meant when you told me that previous recursive course creating approach was better. My bad.
It's up to you whether leave this or roll back to previous approach. 

But I'd rather leave `SELECT LAST_INSERT_ID();` so as not to make extra DB query for created entity.

File: Controllers/TeacherController.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        }
+
+        [HttpPost("courses/create")]
+        public ActionResult<Course> CreateCourse(CourseViewModel courseViewModel)
+        {
+            var course = _dataRepository.CreateCourse(courseViewModel.course);
+            foreach (var category in courseViewModel.CourseCategories)
+            {
+                if (_dataRepository.GetCategory(category.Id) != null)
+                    _dataRepository.InsertToCategoryCourse(course.Id, category.Id);
```
Comment (kravien) [inline Controllers/TeacherController.cs]: As I see it we should stick to one naming convention for create operation.
Create, Post, Insert - we should finally choose the right one and replace all the occurrences.

Or we could just live with the way it is now (:

**Review by kravien — CHANGES_REQUESTED** [review]
Check my comments. 
Overall looks nice.

File: Controllers/TeacherController.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            if (course == null)
+            {
+                return NotFound();
+            }
+            _dataRepository.DeleteСourse(id);
+            return NoContent();
+        }
+
+        [HttpPost("courses/{course_id}/requestToPublish")]
+        public void RequestToPublish(int course_id)
```
Comment (kravien) [inline Controllers/TeacherController.cs]: There also should be validation whether course with `course_id` exists.

File: Controllers/TeacherController.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            _dataRepository.RequestToPublish(course_id);
+        }
+
+        [HttpPost("courses/{course_id}/publish")]
+        public void PublishCourse(int course_id)
```
Comment (kravien) [inline Controllers/TeacherController.cs]: Same as above.

File: Controllers/CourseController.cs [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+            if (course == null)
+            {
+                return NotFound();
+            }
+            return course;
+        }
+
+        [HttpGet("{courseId}/modules")]
+        public IEnumerable<Module> GetModules(int courseId)
```
Comment (kravien) [inline Controllers/CourseController.cs:46]: Missing `course_id` validation.

File: Controllers/CourseController.cs [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            int userId = 5;
+            Course course = _dataRepository.IsEnrolled(userId, courseId) ?
+                _dataRepository.GetCourse(courseId) : null;
+            Module module = course != null ? course.Modules.Where(m => m.Id == moduleId).FirstOrDefault() : null;
+            Lesson lesson = module != null ? module.Lessons.Where(l => l.Id == lessonId).FirstOrDefault() : null;
+            return _dataRepository.GetLessonComponents(lesson.Id);
+        }
+
+        [HttpGet("component/{component_id}/comment")]
+        public IEnumerable<Comment> GetComments(int component_id)
```
Comment (kravien) [inline Controllers/CourseController.cs]: Missing validation again.

**Review by kirsider — CHANGES_REQUESTED** [review]
I've not found bugs anymore...

File: Data/DataRepository.cs (hunk 1 above)
Comment (danikusenko) [reply Data/DataRepository.cs]: Ok, got it.

File: Controllers/TeacherController.cs (hunk 3 above)
Comment (danikusenko) [reply Controllers/TeacherController.cs]: I have changed the name of method "InsertToCategotyCourse". So, the new name is "CreateRecordingForCategoryCourse".

File: Data/DataRepository.cs (hunk 2 above)
Comment (danikusenko) [reply Data/DataRepository.cs:310]: I haven't changed anything in this moment. Let it stay as it is.

File: Controllers/TeacherController.cs (hunk 4 above)
Comment (danikusenko) [reply Controllers/TeacherController.cs]: I have added validation of existing course.

File: Controllers/TeacherController.cs (hunk 5 above)
Comment (danikusenko) [reply Controllers/TeacherController.cs]: Same as above answer.

File: Controllers/CourseController.cs (hunk 6 above)
Comment (danikusenko) [reply Controllers/CourseController.cs:46]: I have added course validation. Method will get modules if we enroll in a course.

File: Controllers/CourseController.cs (hunk 7 above)
Comment (danikusenko) [reply Controllers/CourseController.cs]: I have added validation that starts from course validation and ends with component validation.

File: Controllers/TeacherController.cs (hunk 3 above)
Comment (kravien) [reply Controllers/TeacherController.cs]: Why not just `CreateCategoryCourse`?

File: Controllers/CourseController.cs [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+            int userId = 5;
+            Course course = _dataRepository.IsEnrolled(userId, courseId) ?
+                _dataRepository.GetCourse(courseId) : null;
+            Module module = course != null ? course.Modules.Where(m => m.Id == moduleId).FirstOrDefault() : null;
+            Lesson lesson = module != null ? module.Lessons.Where(l => l.Id == lessonId).FirstOrDefault() : null;
+            return _dataRepository.GetLessonComponents(lesson.Id);
+        }
+
+        [HttpGet("{courseId}/modules/{moduleId}/{lessonId}/component/{component_id}/comment")]
+        public IEnumerable<Comment> GetComments(int courseId, int moduleId, int lessonId, int component_id)
```
Comment (kravien) [inline Controllers/CourseController.cs]: Why would you add this complex route structure? It was okay. I don't think that front-end would send request for component which exists and doesn't belong to any course, module and lesson.

@kirsider, help us out.

File: Controllers/CourseController.cs [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        {
+            int userId = 5;
+            Course course = _dataRepository.IsEnrolled(userId, courseId) ?
+                _dataRepository.GetCourse(courseId) : null;
+            Module module = course != null ? course.Modules.Where(m => m.Id == moduleId).FirstOrDefault() : null;
+            Lesson lesson = module != null ? module.Lessons.Where(l => l.Id == lessonId).FirstOrDefault() : null;
+            Component component = lesson != null ? lesson.Components.Where(c => c.Id == component_id).FirstOrDefault() : null; ; ;
```
Comment (kravien) [inline Controllers/CourseController.cs]: 3 semicolons here.

File: Controllers/TeacherController.cs [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+                return NotFound();
+            }
+            _dataRepository.DeleteСourse(id);
+            return NoContent();
+        }
+
+        [HttpPost("courses/{course_id}/requestToPublish")]
+        public void RequestToPublish(int course_id)
+        {
+            if (_dataRepository.GetBaseCourse(course_id) != null)
```
Comment (kravien) [inline Controllers/TeacherController.cs]: You should add `Not found();` in case course doesn't exist. See other `Get` methods for reference.

**Review by kravien — CHANGES_REQUESTED** [review]
Some more changes needed.

File: Controllers/TeacherController.cs (hunk 10 above)
Comment (danikusenko) [reply Controllers/TeacherController.cs]: Got it

File: Controllers/CourseController.cs (hunk 9 above)
Comment (danikusenko) [reply Controllers/CourseController.cs]: Got it

File: Controllers/CourseController.cs (hunk 8 above)
Comment (danikusenko) [reply Controllers/CourseController.cs]: I dont't know how can I write it in another way.

**Pull request merged**